aws/ec2.go

```diff
@@ -0,0 +1,226 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	awssession "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	log "github.com/sirupsen/logrus"
+
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+)
+
+const (
+	timeout         = 20
+	retries         = 3
+	deviceIndexZero = 0
+)
+
+func convertError(err error) string {
+	if awsErr, ok := err.(awserr.Error); ok {
+		return fmt.Sprintf("%s: %s", awsErr.Code(), awsErr.Message())
+	}
+
+	return fmt.Sprintf("%v", err.Error())
+}
+
+func retriable(err error) bool {
+	if awsErr, ok := err.(awserr.Error); ok {
+		switch awsErr.Code() {
+		case "InternalError":
+			return true
+		case "InternalFailure":
+			return true
+		case "RequestLimitExceeded":
+			return true
+		case "ServiceUnavailable":
+			return true
+		case "Unavailable":
+			return true
+		}
+	}
+
+	return false
+}
+
+func checkSourceDestinationValueIsDisable(check string) bool {
+	return check == apiv3.AWSSrcDstCheckOptionDisable
+}
+
+func UpdateSrcDstCheck(check string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout*time.Second)
+	defer cancel()
+
+	ec2Cli, err := newEC2Client(ctx)
+	if err != nil {
+		return err
+	}
+
+	ec2NetId, err := ec2Cli.getEC2NetworkInterfaceId(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting ec2 network-interface-id: %s", convertError(err))
+	}
+
+	checkVal := !checkSourceDestinationValueIsDisable(check)
+	err = ec2Cli.setEC2SourceDestinationCheck(ctx, ec2NetId, checkVal)
+	if err != nil {
+		return fmt.Errorf("error setting src-dst-check for network-interface-id: %s", convertError(err))
+	}
+
+	log.Infof("Successfully set source-destination-check to %t on network-interface-id: %s", checkVal, ec2NetId)
+	return nil
+}
+
+// Interface for EC2 Metadata service.
+type ec2MetadaAPI interface {
+	AvailableWithContext(ctx aws.Context) bool
+	GetInstanceIdentityDocumentWithContext(ctx aws.Context) (ec2metadata.EC2InstanceIdentityDocument, error)
+	RegionWithContext(ctx aws.Context) (string, error)
+}
+
+func getEC2InstanceID(ctx context.Context, svc ec2MetadaAPI) (string, error) {
+	idDoc, err := svc.GetInstanceIdentityDocumentWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("ec2-instance-id: %s", idDoc.InstanceID)
+	return idDoc.InstanceID, nil
+}
+
+func getEC2Region(ctx context.Context, svc ec2MetadaAPI) (string, error) {
+	region, err := svc.RegionWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("region: %s", region)
+	return region, nil
+}
+
+type ec2Client struct {
+	EC2Svc        ec2iface.EC2API
+	ec2InstanceId string
+}
+
+func newEC2Client(ctx context.Context) (*ec2Client, error) {
+	awsSession, err := awssession.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("error creating AWS session: %v", err)
```

Comment:
Use `%w` with `Errorf()`; it returns a wrapped error so the original error can be accessed if needed.